src/contrib/mlir/lowerer.cpp

```diff
@@ -526,6 +526,58 @@ namespace
         return matchSuccess();
     }
 
+    // Negative
+    REWRITER(NGNegativeOp)
+    {
+        auto loc = cast<NGNegativeOp>(op).getLoc();
+
+        auto result = pass.buildOutputDefs(op, rewriter)[0];
+        NGRAPH_CHECK(result->getType().isa<MemRefType>());
+        // Note that builder's current function is still the original function body.
+        // use getBlock to get the new block instead.
+
+        // get new operands
+        Value* lhs = operands[0];
+
+        ScopedContext scope(rewriter, loc);
+        // Views
+        MemRefView vRes(result), vLHS(lhs);
+        // Index Values
+        IndexedValue iRes(result), iLHS(lhs);
+        // Bounds Index Handles
+        auto lbs = vLHS.getLbs();
+        auto ubs = vLHS.getUbs();
+        // Loop induction vars
+        auto ivs = IndexHandle::makeIndexHandles(vLHS.rank());
+        auto pivs = IndexHandle::makeIndexHandlePointers(ivs);
+        // Steps
+        auto steps = vLHS.getSteps();
+
+        NGRAPH_CHECK(lhs->getType().isa<MemRefType>());
+        Type elemTy = lhs->getType().dyn_cast<MemRefType>().getElementType();
```

Comment:
Since you are checking the type with `isa` in 556, you should use `cast` instead of `dyn_cast`. `cast` will assert if the casting is wrong so we don't have to check that the return is a nullptr.

Reply:
Usually it is  `if(auto elemTy = dyn_cast<MemRefType>(..)) { /* do something with elemTy */ } else { assert(0); }`

Reply:
I've never seen the `dyn_cast` + `assert(0)` idiom. `dyn_cast` is used when multiple kind of objects are expected. For example:

```
if (auto memRefTy = dyn_cast<MemRefType>() {
    do something with memRefTy
} else if (auto tensorTy = dyn_cast<TensorType>() {
    do something with tensorTy
} else if ...
```

If `memRefTy` and `tensorTy` are not needed (do something could be implemented using the base clase), `isa` is used instead, since it's more efficient.

If we expect a single type, such as in this case, `assert(isa<>)` + `cast` or just `cast` is preferred since `cast` has an assert inside.